aws-lc-rs/src/cipher/streaming.rs

```diff
@@ -0,0 +1,527 @@
+    /// If the key is not valid for the cipher algorithm
+    pub fn less_safe_ctr(
+        key: UnboundCipherKey,
+        context: EncryptionContext,
+    ) -> Result<Self, Unspecified> {
+        Self::new(key, OperatingMode::CTR, context)
+    }
+
+    /// CBC cipher mode
+    /// # Errors
+    /// If the key is not valid for the cipher algorithm
+    pub fn cbc_pkcs7(key: UnboundCipherKey) -> Result<Self, Unspecified> {
+        let context = key.algorithm().new_encryption_context(OperatingMode::CBC)?;
+        Self::less_safe_cbc_pkcs7(key, context)
+    }
+
+    /// CBC cipher mode
+    /// # Errors
+    /// If the key is not valid for the cipher algorithm
+    pub fn less_safe_cbc_pkcs7(
+        key: UnboundCipherKey,
+        context: EncryptionContext,
+    ) -> Result<Self, Unspecified> {
+        Self::new(key, OperatingMode::CBC, context)
+    }
+}
+
+/// A cipher decryption key for streaming encryption operations.
+pub struct StreamingDecryptingKey {
+    algorithm: &'static Algorithm,
+    mode: OperatingMode,
+    cipher_ctx: LcPtr<EVP_CIPHER_CTX>,
+}
+impl StreamingDecryptingKey {
+    #[allow(clippy::needless_pass_by_value)]
+    fn new(
+        key: UnboundCipherKey,
+        mode: OperatingMode,
+        context: DecryptionContext,
+    ) -> Result<Self, Unspecified> {
+        let cipher_ctx = LcPtr::new(unsafe { EVP_CIPHER_CTX_new() })?;
+        let algorithm = key.algorithm();
+        let cipher = mode.evp_cipher(key.algorithm);
+        let key_bytes = key.key_bytes.as_ref();
+        debug_assert_eq!(
+            key_bytes.len(),
+            <usize>::try_from(unsafe { EVP_CIPHER_key_length(*cipher) }).unwrap()
+        );
+        let iv = <&[u8]>::try_from(&context)?;
+        debug_assert_eq!(
+            iv.len(),
+            <usize>::try_from(unsafe { EVP_CIPHER_iv_length(*cipher) }).unwrap()
+        );
+
+        if 1 != unsafe {
+            EVP_DecryptInit_ex(
+                cipher_ctx.as_mut_ptr(),
+                *cipher,
+                null_mut(),
+                key_bytes.as_ptr(),
+                iv.as_ptr(),
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        Ok(Self {
+            algorithm,
+            mode,
+            cipher_ctx,
+        })
+    }
+
+    /// Decrypt the input and return the output.
+    /// # Errors
+    /// Returns an error if the output buffer is too small.
+    pub fn update<'a>(
+        &mut self,
+        input: &[u8],
+        output: &'a mut [u8],
+    ) -> Result<&'a [u8], Unspecified> {
+        if output.len() < (input.len() + self.algorithm.block_len) {
+            return Err(Unspecified);
+        }
+
+        let mut outlen: i32 = output.len().try_into()?;
+        let inlen: i32 = input.len().try_into()?;
+        if 1 != unsafe {
+            EVP_DecryptUpdate(
+                self.cipher_ctx.as_mut_ptr(),
+                output.as_mut_ptr(),
+                &mut outlen,
+                input.as_ptr(),
```

Comment:
what if `input` and `output` overlap?

Reply:
> what if input and output overlap?

this is impossible with rust's aliasing rules (at least not without `unsafe`)